qiita_db/util.py

```diff
@@ -2311,3 +2317,476 @@ def send_email(to, subject, body):
+
+    x_data = []
+    y_data = []
+
+    # Populate the lists with data from the dictionary
+    for x, ys in d.items():
+        for y in ys:
+            x_data.append(x)
+            y_data.append(y)
+
+    x_data = np.array(x_data)
+    y_data = np.array(y_data)
+    ax.set_xscale('log')
+    ax.set_yscale('log')
+    ax.set_ylabel(curr)
+    ax.set_xlabel(COL_NAME)
+
+    best_model, options = _resource_allocation_calculate(
+        _df, x_data, y_data, models, curr, COL_NAME, 100)
+    k, a, b = options.x
+    x_plot = np.array(sorted(_df[COL_NAME].unique()))
+    y_plot = best_model(x_plot, k, a, b)
+    ax.plot(x_plot, y_plot, linewidth=1, color='orange')
+
+    maxi = naturalsize(_df[curr].max(), gnu=True) if curr == "MaxRSSRaw" else \
+        timedelta(seconds=float(_df[curr].max()))
+    cmax = naturalsize(max(y_plot), gnu=True) if curr == "MaxRSSRaw" else \
+        timedelta(seconds=float(max(y_plot)))
+
+    mini = naturalsize(_df[curr].min(), gnu=True) if curr == "MaxRSSRaw" else \
+        timedelta(seconds=float(_df[curr].min()))
+    cmin = naturalsize(min(y_plot), gnu=True) if curr == "MaxRSSRaw" else \
+        timedelta(seconds=float(min(y_plot)))
+
+    x_plot = np.array(_df[COL_NAME])
+    failures_df = _resource_allocation_failures(
+        _df, k, a, b, best_model, COL_NAME, curr)
+    failures = failures_df.shape[0]
+
+    ax.scatter(failures_df[COL_NAME], failures_df[curr], color='red', s=3,
+               label="failures")
+
+    ax.set_title(f'{cname}: {sname}\n real: {mini} || {maxi}\n'
+                 f'calculated: {cmin} || {cmax}\n'
+                 f'failures: {failures}')
+    ax.legend(loc='upper left')
+    return best_model, options
+
+
+def _resource_allocation_calculate(_df, x, y, models, type_, COL_NAME, depth):
+    """Helper function for resource allocation plot. Calculates best_model and
+    best_result given the models list and x,y data.
+
+    Parameters
+    ----------
+    x: pandas.Series (pandas column), required
+        Represents x data for the function calculation
+    y: pandas.Series (pandas column), required
+        Represents y data for the function calculation
+    type_: str, required
+        current type (e.g. MaxRSSRaw)
+    COL_NAME: str, required
+        Specifies x axis for the graph
+    models: list, required
+        List of functions that will be used for visualization
+    depth: int, required
+        Maximum number of iterations in binary search
+
+    Returns
+    ----------
+    best_model: function
+        best fitting function for the current list models
+    best_result: object
+        object containing constants for the best model (e.g. k, a, b in kx+b*a)
+    """
+
+    init = [1, 1, 1]
+    best_model = None
+    best_result = None
+    best_failures = np.inf
+    best_max = np.inf
+    counter = 0
+    for model in models:
+        counter += 1
+        # start values for binary search, where sl is left, sr is right
+        sl = 2000
```

Comment:
The comment is appreciated but these still seem like magic numbers to me. A longer description of sl and sr might be nice to have.